### ISSUE TITLE
Adjoint optimization fails for EVEN_Z parity in 2D

### ISSUE THREAD

**Issue by yaugenst** [NanoComp/meep#2438]
This might very well be related to me misunderstanding parities (wouldn't be the first time), so a quick summary of my understanding: When using eigenmode sources, `mp.EVEN_Y + mp.ODD_Z` corresponds to TM polarization with in-plane $H_x$ and $H_y$ components and an out-of-plane $E_z$-component. Conversely, the parity `mp.ODD_Y + mp.EVEN_Z` has the field components $(E_x, E_y, H_z)$, i.e. TE.

For the above TM polarization, adjoint optimization works as expected (and this is also what is used in many of the tutorials), but when I switch to TE, it seems to fail. I have experienced this with different setups and reduced this to a MWE with a simple straight waveguide without any filtering or projection.
![te_vs_tm](https://user-images.githubusercontent.com/1576526/224723441-2b7fff2c-976c-4a1f-b8e2-6c1f68be8fe2.png)

As you can see, the TM case looks like a regular adjoint optimization using MMA, but the TE one goes in the wrong direction and then just stagnates.
I did check if perhaps the gradient sign is wrong in the TE case, but changing the sign doesn't seem to help.
Just to be sure, I also tested this without the y-parities and just using `mp.ODD_Z` for TM and `mp.EVEN_Z` for TE, for which I get the exact same plots.

Is this result expected? Is there something I am misunderstanding about these parities for adjoint optimization?

Below is the code for the plot, using Meep  v1.25 (conda mpich build):
```python
import matplotlib.pyplot as plt
import meep as mp
import meep.adjoint as mpa
import nlopt
import numpy as np

parities = {"TM": mp.EVEN_Y + mp.ODD_Z, "TE": mp.ODD_Y + mp.EVEN_Z}
# parities = {"TM": mp.ODD_Z, "TE": mp.EVEN_Z}

resolution = 20
max_eval = 30
cell = mp.Vector3(6, 4)
design_region_size = (2, 1)
nx = design_region_size[0] * resolution
ny = design_region_size[1] * resolution
fcen = 1 / 1.55
df = 0.2
dpml = 1.0

mp.verbosity.meep = 0
mp.verbosity.mpb = 0

design_region = mpa.DesignRegion(
    mp.MaterialGrid(mp.Vector3(nx, ny), mp.Medium(index=1), mp.Medium(index=2)),
    volume=mp.Volume(center=mp.Vector3(), size=mp.Vector3(*design_region_size)),
)

geometry = [
    mp.Block(
        center=design_region.center,
        size=design_region.size,
        material=design_region.design_parameters,
    ),
    mp.Block(
        center=mp.Vector3(-cell.x / 2 + dpml),
        size=mp.Vector3(2, 0.5),
        material=mp.Medium(index=2),
    ),
    mp.Block(
        center=mp.Vector3(cell.x / 2 - dpml),
        size=mp.Vector3(2, 0.5),
        material=mp.Medium(index=2),
    ),
]

fig, ax = plt.subplots(1, 2, figsize=(8, 4))
ax[1].set_xlabel("iteration")
ax[1].set_ylabel("objective")

for name, parity in parities.items():
    sources = [
        mp.EigenModeSource(
            mp.GaussianSource(frequency=fcen, fwidth=df),
            eig_band=1,
            direction=mp.NO_DIRECTION,
            eig_kpoint=mp.Vector3(1, 0),
            size=mp.Vector3(0, cell.y),
            center=mp.Vector3(-cell.x / 2 + dpml, 0),
            eig_parity=parity,
        )
    ]

    sim = mp.Simulation(
        resolution=resolution,
        cell_size=cell,
        sources=sources,
        geometry=geometry,
        boundary_layers=[mp.PML(dpml)],
        k_point=mp.Vector3(),
    )

    design_region.update_design_parameters(np.full(nx * ny, 0.5))

    m12 = mpa.EigenmodeCoefficient(
        sim,
        volume=mp.Volume(
            center=mp.Vector3(cell.x / 2 - dpml, 0), size=mp.Vector3(0, cell.y)
        ),
        mode=1,
        eig_parity=parity,
    )

    mpa_opt = mpa.OptimizationProblem(
        simulation=sim,
        objective_functions=lambda s12: np.sum(np.abs(s12) ** 2),
        objective_arguments=[m12],
        design_regions=[design_region],
        frequencies=[fcen],
    )

    mpa_opt.plot2D(True, ax=ax[0])

    hist = []

    def nlopt_obj(x, gd):
        f0, meep_grad = mpa_opt([x])
        if gd.size > 0:
            gd[:] = meep_grad
        hist.append(f0)
        return np.real(f0)

    n = nx * ny
    x0 = np.full(n, 0.5)
    lb = np.zeros_like(x0)
    ub = np.ones_like(x0)

    opt = nlopt.opt(nlopt.LD_MMA, n)
    opt.set_lower_bounds(lb)
    opt.set_upper_bounds(ub)
    opt.set_max_objective(nlopt_obj)
    opt.set_maxeval(max_eval)

    opt.optimize(x0)

    ax[1].plot(hist, label=name)

ax[1].legend()
fig.tight_layout()
fig.savefig("te_vs_tm.png")
```

**Comment by mrbaozi** [comment]
Yes usually I normalize the flux, but it doesn't change the behavior of the optimization which is why I omitted the normalization for the MWE.

**Comment by mrbaozi** [comment]
I agree with you in general, but this does not seem to be the case here. I should have added a plot of the optimized structures to the original post.

This is the structure after 30 iterations of optimization in TM:
![TM_optimized](https://user-images.githubusercontent.com/1576526/224739375-45a5df2d-4e9b-42a2-8207-62423e03252c.png)

And the same for TE:
![TE_optimized](https://user-images.githubusercontent.com/1576526/224739479-6b4786e5-fc28-42a7-9738-eccb76d356d3.png)

As you can see, the optimizer really doesn't do anything.

**Comment by smartalecH** [comment]
Can you plot the FOM evolutions for both cases, along with some gradients?

**Comment by mrbaozi** [comment]
Hey @smartalecH, thanks for the quick response! And thanks for clarifying that I got the polarizations backwards, I'll keep that in mind in the future. For this issue I will stick to what I defined originally, just to avoid confusion.

I agree with all your points about normalization and and optimization, but I have excluded those as possibilities before submitting the issue, I just wanted to boil down the MWE to an absolute minimum.

Your hint about the subpixel smoothing seems to be right on point - it fixes the issue.

Below is a plot of the gradients and design variables for the first five iterations of the optimization in the initial post (FOM evolutions are also the same) using material averaging:
![te_vs_tm_grads](https://user-images.githubusercontent.com/1576526/224914193-a221ea75-d269-4760-9eb0-2466c6ce3865.png)

And this is what I get with averaging turned off:
![te_vs_tm_grads_no_avg](https://user-images.githubusercontent.com/1576526/224914255-bfce818c-04df-453b-a196-da3c2c865898.png)

**Comment by smartalecH** [comment]
We can use this as a simple test case when we debug the subpixel smoothing (eventually). 

(cc @oskooi i think it's time to disable smoothing in material grids by default...)
